libraries/AP_ExternalAHRS/AP_ExternalAHRS_VectorNav.cpp

```diff
@@ -371,57 +368,67 @@ bool AP_ExternalAHRS_VectorNav::decode(char c)
 }
 
 // decode the most recently consumed term
-// returns true if new sentence has just passed checksum test and is validated
+// returns true if new term is valid
 bool AP_ExternalAHRS_VectorNav::decode_latest_term()
 {
+    // Check the first two terms (In most cases header + reg number) that they match the sent
+    // message. If not, the response is invalid.
     switch (nmea.term_number) {
         case 0:
-            if (strcmp(nmea.term, "VNRRG") != 0) {
+            if (strncmp(nmea.term, message_to_send, nmea.term_offset) !=
+                0) {  // ignore sync byte in comparison
                 return false;
+            } else if (strncmp(nmea.term, "VNERR", nmea.term_offset) == 0) {
+                AP_HAL::panic("VectorNav received VN error response");
```

Comment:
Panicing is not a great idea, it effectively just gives up. Would you expect to see this in normal operation? In which case it should trigger a pre-arm. Or is it only if there is some edge case and very unlikely? I which case you could throw a internal error.  

Would it be worth trying a reset of the vector nav to see if we could recover?

Reply:
@IamPete1 

The only cases I'd expect to see an error in response is 

1. If something is catastrophically wrong with the VN sensor (hardware failure, etc)
2. If the code was incorrectly edited and sent a command that isn't acceptable to the sensor
3. If the AP-configured IMU Rate (via get_rate()) is set too high, we will reply with insufficient baudrate VNERR

I don't think any of those cases would be resolved by resetting the sensor, but I agree especially since (3) is very viable with normal use panicing may not be the not the correct response. Given the failure cases, what would you suggest instead?

Reply:
OK, in that case, I would print the error specific (`GCS_SEND_TEXT`) so the user will have some feedback. Then you need to make sure that they can't arm. Hopefully the existing health checks are sufficient, if not then update those so they give a "init failed" error. 

Really there are three things that you need to acheave:
- Not prevent the rest of the code running. 
- Give the user a clue so they can track down how to resolve
- Prevent the vehicle flying in a unsafe state

Reply:
Just pushed a change to resolve this issue.

In this change, the system will continue to send the message to the VectorNav every READ_REQUEST_RETRY_MS (500ms). The pre-arm check should continue to show as a failure, because the system will never pass initialization (assuming the VNERR is never resolved) THis is consistent with previous behavior: block initialization and retry until a valid response is received. It will result in the received error message being sent to GCS every 500ms.